src/sketches.py

```diff
@@ -0,0 +1,366 @@
+from __future__ import annotations
+from .schema import (
+    DerivedTypeVariable,
+    Lattice,
+    FreshVarFactory,
+    ConstraintSet,
+    SubtypeConstraint,
+    Variance,
+    RetypdError,
+)
+from .loggable import Loggable, LogLevel
+import os
+import networkx
+from typing import Set, Union, Optional, Tuple, Dict
+
+
+class SketchNode:
+    """
+    A node in a sketch graph. The node is associated with a DTV and it
+    captures its upper and lower bound in the type lattice.
+    A sketch node might be referenced by `LabelNode` in recursive types.
+    If that is the case, the sketch node can represent the primitive
+    type infinite DTVs, e.g.:
+    f.in_0, f.in_0.load.σ4@0, f.in_0.load.σ4@0.load.σ4@0, ...
+    """
+
+    def __init__(
+        self,
+        dtv: DerivedTypeVariable,
+        lower_bound: DerivedTypeVariable,
+        upper_bound: DerivedTypeVariable,
+    ) -> None:
+        self._dtv = dtv
+        self.lower_bound = lower_bound
+        self.upper_bound = upper_bound
+        # Reference to SketchNodes (in other SCCs) that this node came from
+        self.source: Set[SketchNode] = set()
+        self._hash = hash(self._dtv)
+
+    @property
+    def dtv(self):
+        """
+        The main DTV represented by the sketch node.
+        """
+        return self._dtv
+
+    @dtv.setter
+    def dtv(self, value):
+        raise NotImplementedError("Read-only property")
+
+    # the atomic type of a DTV is an annotation, not part of its identity
+    def __eq__(self, other) -> bool:
+        if isinstance(other, SketchNode):
+            return self.dtv == other.dtv
+        return False
+
+    def __hash__(self) -> int:
+        return self._hash
+
+    def __str__(self) -> str:
+        return f"({self.lower_bound} <= {self.dtv} <= {self.upper_bound})"
```

Comment:
I don't have strong feelings as long as we're consistent, but I'm curious about the change from `'` to `"`.

Reply:
This is done by `black` the code formatting tool https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#strings